common/flatpak-prune.c

```diff
@@ -0,0 +1,813 @@
+      bag->current_chunk_used = 0;
+      bag->chunks = g_slist_prepend (bag->chunks, bag->current_chunk);
+    }
+
+  res = &bag->current_chunk[bag->current_chunk_used++];
+  memcpy (res, name, sizeof (FlatpakOstreeObjectName));
+
+  if (bag->current_chunk_used == BAG_CHUNK_SIZE)
+    bag->current_chunk = NULL; /* Need new chunk */
+
+  g_hash_table_add (bag->hash, res);
+}
+
+/* Find all reachable commit objects starting from any ref in the repo
+ * optionally limiting the number of parent commits.
+ *
+ * This doesn't do any locking, so need something else to have an exclusive lock
+ * on the repo to avoid races with other processes modifying the repo.
+ */
+static gboolean
+traverse_reachable_refs_unlocked (OstreeRepo  *repo,
+                                  guint        depth,
+                                  FlatpakOstreeObjectNameBag *reachable,
+                                  GCancellable *cancellable,
+                                  GError      **error)
+{
+  g_autoptr(GHashTable) all_refs = NULL;  /* (element-type utf8 utf8) */
+  g_autoptr(GHashTable) all_collection_refs = NULL;  /* (element-type OstreeChecksumRef utf8) */
+  g_autoptr(GHashTable) checksums = NULL;  /* (element-type const char *) */
+
+  checksums = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  /* Get all commits up to depth from the regular refs */
+  if (!ostree_repo_list_refs (repo, NULL, &all_refs,
+                              cancellable, error))
+    return FALSE;
+
+  GLNX_HASH_TABLE_FOREACH_V (all_refs, const char*, checksum)
+    {
+      if (!traverse_commit_parents_unlocked (repo, checksum, depth, checksums, cancellable, error))
+        return FALSE;
+    }
+
+  /* Get all commits up to depth from the collection refs */
+  if (!ostree_repo_list_collection_refs (repo, NULL, &all_collection_refs,
+                                         OSTREE_REPO_LIST_REFS_EXT_EXCLUDE_REMOTES, cancellable, error))
+    return FALSE;
+
+  GLNX_HASH_TABLE_FOREACH_V (all_collection_refs, const char*, checksum)
+    {
+      if (!traverse_commit_parents_unlocked (repo, checksum, depth, checksums, cancellable, error))
+        return FALSE;
+    }
+
+  /* Find reachable objects from each commit checksum */
+  GLNX_HASH_TABLE_FOREACH_V (checksums, const char*, checksum)
+    {
+      g_autoptr(GVariant) extra_commitmeta = NULL;
+      g_autoptr(GVariant) commit_reachable = NULL;
+      FlatpakOstreeObjectName commit_name;
+
+      /* Early bail-out if we already scanned this commit in the first phase (or via some other branch) */
+      flatpak_ostree_object_name_serialize (&commit_name, checksum, OSTREE_OBJECT_TYPE_COMMIT);
+      if (object_name_bag_contains (reachable, &commit_name))
+        continue;
+
+      flatpak_debug2 ("Finding objects to keep for commit %s", checksum);
+
+      if (!load_extra_commitmeta (repo, checksum, &extra_commitmeta, cancellable, error))
```

Comment:
Do we need to check if the commit is partial and avoid the `commitmeta2` caching if so? Because if the commit is partial but then it later becomes complete wouldn't the `xa.reachable` info become outdated and no longer correct? The docs for `ostree_repo_mark_commit_partial` say:
```
 * Commits in the "partial" state do not have all their child objects
 * written.  This occurs in various situations, such as during a pull,
 * but also if a "subpath" pull is used, as well as "commit only"
 * pulls.
```

And I guess the server side repo shouldn't normally have subpath commits like clients do for locale extensions, but maybe it comes up in some situation.

Reply:
Its true that its unlikely to happen in the way this is regularly used, but we should still handle it.

Reply:
I think its fine reading it actually. However, we should never save a new one if what we generated is based on a partial commit.

Reply:
Gated the commitmeta2 save on on-partial state.